src/Microsoft.FeatureManagement/Telemetry/TelemetryEventHandler.cs

```diff
@@ -0,0 +1,92 @@
+﻿using Microsoft.Extensions.Logging;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+
+namespace Microsoft.FeatureManagement.Telemetry
+{
+    internal static class TelemetryEventHandler
+    {
+        private static readonly string EvaluationEventVersion = "1.0.0";
+
+        /// <summary>
+        /// Handles an evaluation event by adding it as an activity event to the current Activity.
+        /// </summary>
+        /// <param name="evaluationEvent">The <see cref="EvaluationEvent"/> to publish as an <see cref="ActivityEvent"/></param>
+        /// <param name="logger">Optional logger to log warnings to</param>
+        public static void HandleEvaluationEvent(EvaluationEvent evaluationEvent, ILogger logger)
+        {
+            Debug.Assert(evaluationEvent != null);
+            Debug.Assert(evaluationEvent.FeatureDefinition != null);
+
+            var tags = new ActivityTagsCollection()
+            {
+                { "FeatureName", evaluationEvent.FeatureDefinition.Name },
+                { "Enabled", evaluationEvent.Enabled },
+                { "VariantAssignmentReason", evaluationEvent.VariantAssignmentReason },
+                { "Version", EvaluationEventVersion }
+            };
+
+            if (!string.IsNullOrEmpty(evaluationEvent.TargetingContext?.UserId))
+            {
+                tags["TargetingId"] = evaluationEvent.TargetingContext.UserId;
+            }
+
+            if (!string.IsNullOrEmpty(evaluationEvent.Variant?.Name))
+            {
+                tags["Variant"] = evaluationEvent.Variant.Name;
+            }
+
+            if (evaluationEvent.FeatureDefinition.Telemetry.Metadata != null)
+            {
+                foreach (KeyValuePair<string, string> kvp in evaluationEvent.FeatureDefinition.Telemetry.Metadata)
+                {
+                    if (tags.ContainsKey(kvp.Key))
+                    {
+                        logger?.LogWarning($"{kvp.Key} from telemetry metadata will be ignored, as it would override an existing key.");
+
+                        continue;
+                    }
+
+                    tags[kvp.Key] = kvp.Value;
```

Comment:
Can we instead emit all this metadata as an object with key `Metadata` ?

Reply:
We could, but I don't think it adds any value. This is the outgoing event that will go to App Insights. As far as their concerned- this should just flattened data. These fields are only considered `metadata` because they were given to FM from an outside source (like the provider).